### PR TITLE
Introduce architecture subdirectory for Windows build (#129805, #116196)

### PR DESCRIPTION
**Pull request by p-b-o**
To implement windows-arm64 support, it is needed to add architecture as a subdirectory (https://github.com/flutter/flutter/issues/129805).

In short, when performing a flutter windows build, we have:
- Before: build/windows/runner/Release/gallery.exe
- After: build/windows/x64/runner/Release/gallery.exe

This convention follows what flutter linux build does.

The old "runner" folder is automatically cleaned to prevent confusion for users not aware of this change, and who could continue to use outdated binaries accidentally.

We didn't introduce any change in ephemeral build folder. This might be required when introducing another architecture, but it won't be a breaking change, as those files are not used by end user.

Since we use a fresh cmake folder, we introduce FLUTTER_TARGET_PLATFORM in flutter cmake file (https://github.com/flutter/flutter/issues/116196). This will be needed when targeting another architecture than x64, and will avoid to have to clean existing cmake files. An automatic migration was implemented to automate this change in existing applications.

Finally, we set a specific architecture when calling cmake. This fixes existing build and tests on Windows on Arm, where cmake tries to compile native windows-arm64 binaries by default (which fails to link x64 libraries). In more, it will avoid need to clean existing cmake files when introducing another architecture.

Design doc: [flutter.dev/go/windows-arm64](https://flutter.dev/go/windows-arm64)

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by pbo-linaro** [comment]
@loic-sharma Here it is!
Your comments are welcome, and those from other flutter devs as well.

For now, I focused on a minimal change, with tests working (on x64 and arm64).
Probably I can update some documentation as well, don't hesitate to say.

**Comment by pbo-linaro** [comment]
Fixed error reported by flutter analyze.

**Comment by pbo-linaro** [comment]
@loic-sharma I'm not sure if the failure (windows plugin_test) means that there is something missing, or if the test script itself (located elsewhere?) should be updated to reflect new path to binaries. If you could help on this, that would be nice :).

**Comment by andrewkolos** [comment]
@loic-sharma Would you be interested in reviewing this?

**Comment by loic-sharma** [comment]
Yup will do!

File: packages/flutter_tools/lib/src/build_info.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/build_info.dart:877]: Should we use `x64` instead of the host platform's architecture since for now we'll build an x64 binary on Windows Arm64?

File: packages/flutter_tools/lib/src/build_info.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/build_info.dart:880]: Can we simplify here?

```suggestion
  return globals.fs.path.join(getBuildDirectory(), 'windows', arch);
```

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:194]: Can we move this to the generated configs in `_writeGeneratedFlutterConfig`?

https://github.com/flutter/flutter/blob/b7acafa0bbf6a23943c447eb98a6b21b1adcbe93/packages/flutter_tools/lib/src/cmake.dart#L101-L109

File: packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart:46]:
```suggestion
    // Skip this migration if the affected file does not exist. This indicates
```

File: packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart:38]: Please also update the app template so that new projects don't need to be migrated:

https://github.com/flutter/flutter/blob/b7acafa0bbf6a23943c447eb98a6b21b1adcbe93/packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt#L81-L97

File: packages/flutter_tools/test/general.shard/windows/migrations/build_architecture_migration_test.dart
Comment (loic-sharma) [inline packages/flutter_tools/test/general.shard/windows/migrations/build_architecture_migration_test.dart:125]:
```suggestion
    testWithoutContext('migrates project to set the target platform', () {
```

File: packages/flutter_tools/test/general.shard/windows/migrations/build_architecture_migration_test.dart
Comment (loic-sharma) [inline packages/flutter_tools/test/general.shard/windows/migrations/build_architecture_migration_test.dart:162]:
```suggestion
    testWithoutContext('migrates project to set the target platform (CRLF)', () {
```

File: packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart:32]: We should make sure to fallback to `windows-x64` if `FLUTTER_TARGET_PLATFORM` is not set.

Imagine the user updates to the latest Flutter SDK and then reverts to an older SDK. The `windows/flutter/CMakeLists.txt` file will have been migrated to use `FLUTTER_TARGET_PLATFORM` but the older Flutter SDK won't set `FLUTTER_TARGET_PLATFORM`. As a result, the user's build is broken until they revert the migration to `windows/flutter/CMakeLists.txt`. This would be particularly problematic while this change is in some but not all release channels (switching between release channels would break your build).

**Comment by loic-sharma** [comment]
> @loic-sharma I'm not sure if the failure (windows plugin_test) means that there is something missing, or if the test script itself (located elsewhere?) should be updated to reflect [new path to binary plugintest_test.exe](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8773783338479745633/+/u/run_plugin_test_windows/test_stdout).

It looks like you'll need to update this:

https://github.com/flutter/flutter/blob/53082f65b57a5f008d5b6c1d48510f5450b4e246/dev/devicelab/lib/tasks/plugin_tests.dart#L296-L300

You should be able to run this test locally using these instructions: https://github.com/flutter/flutter/tree/master/dev/devicelab#running-tests-locally

TLDR, this should work:

```ps1
cd path/to/flutter
cd dev/devicelab
../../bin/cache/dart-sdk/bin/dart bin/test_runner.dart test -t plugin_test_windows
```

Let me know if you run into any problems!

**Comment by pbo-linaro** [comment]
First fixing failed test, before addressing other comments in review.

File: packages/flutter_tools/lib/src/build_info.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/build_info.dart:877]: For now, windows-arm64 does not exist on Flutter, so it's x64 that will be chosen:
- https://github.com/flutter/flutter/blob/58332e95d66c784220ddb3a5d397b598c3aa03cc/packages/flutter_tools/lib/src/build_info.dart#L807

For now, I would prefer to leave it like this, else it's needed to add this target parameter everywhere...

If this does not satisfy you, I can simply hardcode 'x64'. My goal is to keep this PR simple and focused on a simple change (not to introduce the possibility of a new potential targetPlatform).

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:194]: Linux build adds it directly to the cmake command, and IMHO, it's more explicit that putting it in another function.
FLUTTER_VERSION_* stuff is the same for all architecture, thus it makes sense to factorize it in a single place.

What do you think?

**Comment by pbo-linaro** [comment]
Pushed resolved changes.

**Comment by pbo-linaro** [comment]
Fixed analyze not happy with raw cmake strings 'if(...'.

File: packages/flutter_tools/lib/src/build_info.dart
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/build_info.dart:877]: > For now, windows-arm64 does not exist on Flutter, so it's x64 that will be chosen:

Ah gotcha, I didn't notice that.

> I can simply hardcode 'x64'. My goal is to keep this PR simple and focused on a simple change (not to introduce the possibility of a new potential targetPlatform).

Yup, hardcoding seems perfectly reasonable to me.

It looks like `getWindowsBuildDirectory` is only used in two places:

https://github.com/flutter/flutter/blob/9cc4f94397f1f4b4d49ff38a02e5a0708dc9d68e/packages/flutter_tools/lib/src/windows/build_windows.dart#L82

https://github.com/flutter/flutter/blob/9cc4f94397f1f4b4d49ff38a02e5a0708dc9d68e/packages/flutter_tools/lib/src/windows/application_package.dart#L114

I'd lean towards making the argument required and updating those two call sites to hard code `TargetPlatform.windows_x64`. That makes it a bit clearer what's happening from the call sites.

File: packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt
Comment (loic-sharma) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt:90]: This would likely be better near the top of the file in case we ever need to re-use this variable again for something else. Perhaps something like:

```diff
  # TODO: Move the rest of this into files in ephemeral. See
  # https://github.com/flutter/flutter/issues/57146.
+
+ # Set fallback configurations for older versions of the flutter tool.
  if (NOT DEFINED FLUTTER_TARGET_PLATFORM)
    set(FLUTTER_TARGET_PLATFORM "windows-x64")
  endif()
+
  set(WRAPPER_ROOT "${EPHEMERAL_DIR}/cpp_client_wrapper")
```

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:194]: Hm I agree we should align with Linux. If we move `FLUTTER_TARGET_PLATFORM` to the generated CMake config file, we would need to also update Linux. That seems out-of-scope for this PR, and we can follow-up with this change later if needed.

That said, I'm curious as to why Linux chose this approach. The generated cmake config file makes it easier to use the CMake project without the Flutter tool. Regardless, I'll close this off!

File: packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt
Comment (pbo-linaro) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt:90]: I agree, but this make the migration change more painful to implement. Could we please simply start like this, and update the day someone else will need this variable? It would be hard to miss this using a simple grep.

Another point: For Linux, this variable is not needed anywhere else, so I doubt it will be needed on Windows.

File: packages/flutter_tools/lib/src/build_info.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/build_info.dart:877]: Will do it 👍

**Comment by pbo-linaro** [comment]
Added change to hardcode TargetPlatform to getWindowsBuildDirectory.

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:196]: Thanks for accepting this.

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:86]:
```suggestion
  final Directory buildDirectory = globals.fs.directory(
    getWindowsBuildDirectory(targetPlatform),
  );
```

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:144]:
```suggestion
        globals.fs.path.join(
          getWindowsBuildDirectory(TargetPlatform.windows_x64),
          'runner',
          'Release',
        ),
```

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:115]: Could we move this logic into `BuildArchitectureMigration`? This would have the benefit of cleaning up the build folder before we run the build, which may help prevent weird issues.

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:114]: Ideally this should have the same logic as the `clean` command's deletion:

https://github.com/flutter/flutter/blob/d9cb50e63de79aa71caa97ecb50f8d06596b3aa0/packages/flutter_tools/lib/src/commands/clean.dart#L116-L143

Namely:

1. Show a progress bar
2. If a file in the build directory is in use, show a helpful error message describing how to stop the program using the file

@christopherfujino How would you feel about moving `CleanCommand.deleteFile` to [`FileSystemUtils`](https://github.com/flutter/flutter/blob/d9cb50e63de79aa71caa97ecb50f8d06596b3aa0/packages/flutter_tools/lib/src/base/file_system.dart#L31C7-L31C23)?

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:114]: Hmm, this seems to do a lot more than I would expect a generic method named `deleteFile`. I'm ok with moving htis to FileSystemUtils if we use a more descriptive method name, like `recursivelyDeleteEntity({bool showProgress = false})`

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:115]: Yes, that makes perfect sense!

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:183]: Will update it.

File: packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt
Comment (pbo-linaro) [reply packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt:90]: Yes you're right, it would be more complicated to update later. I'll update then!

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:114]: After trying to move this, it makes several changes that are not so welcome on this PR. (tests moving, etc).
Would that be acceptable to first use the initial version, and then refactor this in a subsequent PR?

**Comment by pbo-linaro** [comment]
Pushed fix for latest comments (cmake file, nit). I left open the discussion about `deleteFile` method, that should be moved in another PR IMHO.

FYI, I'll be off next week, so there will be some delay before I can answer.

File: packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart:80]:
```suggestion
Deleting previous build folder ${oldRunnerDirectory.path}.
```

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:135]:
```suggestion
          buildDirectory,
```

File: packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/migrations/build_architecture_migration.dart:83]: Let's still catch the exception here. The previous build directory contains executables the user might have run, so it's not unlikely this will throw:

```suggestion
      try {
        oldRunnerDirectory.deleteSync(recursive: true);
      } on FileSystemException catch (error) {
        final String path = oldRunnerDirectory.path;
        globals.printError(
          'Failed to remove $path. '
          'A program may still be using a file in the directory or the directory itself. '
          'To find and stop such a program, see: '
          'https://superuser.com/questions/1333118/cant-delete-empty-folder-because-it-is-used'
        );
      }
```

File: packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt
Comment (loic-sharma) [inline packages/flutter_tools/templates/app_shared/windows.tmpl/flutter/CMakeLists.txt:92]: Migrating older versions of this file should result in the exact same file as the latest template. The migration doesn't add a new line here. We should either remove the new line here, or, update the migration to insert this new line:

```suggestion
```

**Comment by loic-sharma** [comment]
@pbo-linaro The `flutter run` devicelab test needs to be updated to match the latest build directory:

https://github.com/flutter/flutter/blob/83f56ee77f7cd54baffc4bbe8fea73c441094a38/dev/devicelab/lib/tasks/run_tests.dart#L180-L182

https://github.com/flutter/flutter/blob/83f56ee77f7cd54baffc4bbe8fea73c441094a38/dev/devicelab/lib/tasks/run_tests.dart#L208

**Comment by loic-sharma** [comment]
@pbo-linaro Also, as part of this pull request please migrate [all of the `windows/flutter/CMakeLists.txt` files in this repository](https://github.com/search?q=repo%3Aflutter%2Fflutter+windows-x64+path%3A**windows%2Fflutter%2FCMakeLists.txt&type=code) to use the `FLUTTER_TARGET_PLATFORM`.

**Comment by pbo-linaro** [comment]
Just pushed a new version with all suggested changes:
- All CMakeFiles migration
- fixes for log messages
- fix for CMakeFile extra line
- Catch exception when deleting old runner directory
- Fixes for device lab tests

File: examples/platform_view/windows/flutter/CMakeLists.txt
Comment (loic-sharma) [inline examples/platform_view/windows/flutter/CMakeLists.txt:4]: Please keep the license header if it exists

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (loic-sharma) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:171]:
```suggestion
  if (targetPlatform != TargetPlatform.windows_x64) {
    throwToolExit('Windows build supports only x64 target architecture');
  }
```

**Review by loic-sharma — COMMENTED** [review]
These changes look good to me, but I'd like to run some more manual tests before approving. Thank you for the **excellent** contribution!

**Review by yaakovschectman — APPROVED** [review]
Looks okay to me, pending waiting comments or approval from Loic, and keeping the license headers where applicable.

File: dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt
Comment (cbracken) [inline dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt:4]: Any reason for the copyright header removal?

File: examples/flutter_view/windows/flutter/CMakeLists.txt
Comment (cbracken) [inline examples/flutter_view/windows/flutter/CMakeLists.txt:4]: Similar here.

**Review by cbracken — COMMENTED** [review]
Overall looks great! From my end, just a couple nits re: three copyright headers removed.

File: dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt
Comment (christopherfujino) [reply dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt:4]: yeah, requesting changes on this so we don't accidentally merge this

**Review by christopherfujino — CHANGES_REQUESTED** [review]

File: dev/a11y_assessments/windows/flutter/CMakeLists.txt
Comment (christopherfujino) [inline dev/a11y_assessments/windows/flutter/CMakeLists.txt:13]: @loic-sharma how long do you think we would need to keep this here? Not having this would only affect users in the situation where an app was created with Flutter version N, and then someone tried to build it from Flutter version N - 1 or earlier, right?

Comment (christopherfujino) [reply dev/a11y_assessments/windows/flutter/CMakeLists.txt:13]: oh wait, this is a test. Do we even need this?

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/windows/build_windows.dart:170]:
```suggestion
    throwToolExit('windows build is only supported for windows-x64');
```

Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:170]: oh nvm, Loic's suggestion is better.

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:191]: What happens if someone directly invokes a build via cmake without providing `-DFLUTTER_TARGET_PLATFORM`?

File: dev/a11y_assessments/windows/flutter/CMakeLists.txt
Comment (christopherfujino) [reply dev/a11y_assessments/windows/flutter/CMakeLists.txt:13]: Ahh, I see the context from https://github.com/flutter/flutter/pull/131843/files#r1284879965. Were these test apps run with the migrator?

File: examples/platform_view/windows/flutter/CMakeLists.txt
Comment (pbo-linaro) [reply examples/platform_view/windows/flutter/CMakeLists.txt]: I just copied the template, and indeed, it's not fully coherent between all the copies of this file.

**Comment by pbo-linaro** [comment]
> These changes look good to me, but I'd like to run some more manual tests before approving. Thank you for the **excellent** contribution!

Thanks for your very kind help on this, and the design document.

File: dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt
Comment (pbo-linaro) [reply dev/benchmarks/complex_layout/windows/flutter/CMakeLists.txt]: I simply copied the template file to replace all its copies, and it seems like copyright is present on some, and not on others. For now, I just reintegrated copyright where it was, and we can fix this later.

File: packages/flutter_tools/lib/src/windows/build_windows.dart
Comment (pbo-linaro) [reply packages/flutter_tools/lib/src/windows/build_windows.dart:57]: Added a TODO with correct issue.

**Comment by pbo-linaro** [comment]
- Reintroduced copyright for cmake files having one
- Fix an error message
- Added TODO for targetPlatform

**Comment by pbo-linaro** [comment]
Thanks @yaakovschectman, @christopherfujino and @cbracken for your reviews.

**Review by christopherfujino — APPROVED** [review]
LGTM, thanks so much for this massive contribution!

**Review by loic-sharma — APPROVED** [review]
I ran the following manual test on my machine:

1. Create and run app before this change
2. Update to this change. Run app and verify the app is migrated to a target-specific build
3. Rollback this change. Run app and verify migrated app still works.

I also ran the following Devicelab tests:

Task | x86_64 | arm64 | Notes
-- | -- | -- | --
platform_channel_sample_test_windows | ✅ | ✅ |
plugin_test_windows | ✅ | ✅ | Arm64 machine required turning on developer mode
run_debug_test_windows | ✅ | ✅ |
run_release_test_windows | ✅ | ✅ |
hello_world_win_desktop__compile | ✅ | ✅ | 
flutter_gallery_win_desktop__compile | ✅ | ⚠️ | Arm64 test passes but requires disabling git's SSL verification
flutter_gallery_win_desktop__start_up | ✅ | ✅ |
complex_layout_win_desktop__start_up | ✅ | ✅ |
platform_view_win_desktop__start_up | ✅ | ✅ |
windows_startup_test | ✅ | ✅ |
flutter_tool_startup | ✅ | ❌ | On Arm64, `flutter update-packages --ci` fails with exit code 69. We should investigate this after this PR lands

Excellent work @pbo-linaro! I will land this tomorrow morning so that I can track postsubmit and revert this if there's failures.

**Comment by pbo-linaro** [comment]
Thanks very much for taking the time to review and test this @loic-sharma. Hope we won't run into unexpected use cases.

**Comment by pbo-linaro** [comment]
I ran `flutter update-packages --ci` on flutter repo itself, and gallery app, on windows-arm64 machine, and it worked on both.
Don't hesitate to create an issue for this, and ping me on this one, so I can investigate.

**Pull request merged**
